codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ServiceGenerator.java

```diff
@@ -0,0 +1,136 @@
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen;
+
+import java.util.List;
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.go.codegen.integration.RuntimeClientPlugin;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.traits.StringTrait;
+import software.amazon.smithy.model.traits.TitleTrait;
+
+/**
+ * Generates a service client and configuration.
+ */
+final class ServiceGenerator implements Runnable {
+
+    public static final String CONFIG_NAME = "Options";
+    public static final String API_OPTIONS_FUNC_NAME = "APIOptionFunc";
+
+    private final GoSettings settings;
+    private final Model model;
+    private final SymbolProvider symbolProvider;
+    private final GoWriter writer;
+    private final ServiceShape service;
+    private final List<GoIntegration> integrations;
+    private final List<RuntimeClientPlugin> runtimePlugins;
+
+    ServiceGenerator(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            ServiceShape service,
+            List<GoIntegration> integrations,
+            List<RuntimeClientPlugin> runtimePlugins
+    ) {
+        this.settings = settings;
+        this.model = model;
+        this.symbolProvider = symbolProvider;
+        this.writer = writer;
+        this.service = service;
+        this.integrations = integrations;
+        this.runtimePlugins = runtimePlugins;
+    }
+
+    @Override
+    public void run() {
+        Symbol serviceSymbol = symbolProvider.toSymbol(service);
+        writer.writeShapeDocs(service);
+        writer.openBlock("type $T struct {", "}", serviceSymbol, () -> {
+            writer.write("options $L", CONFIG_NAME);
+        });
+
+        generateConstructor(serviceSymbol);
+
+        String clientId = CodegenUtils.getDefaultPackageImportName(settings.getModuleName());
+        String clientTitle = service.getTrait(TitleTrait.class).map(StringTrait::getValue).orElse(clientId);
+
+        writer.writeDocs("ServiceID returns the name of the identifier for the service API.");
+        writer.write("func (c $P) ServiceName() string { return $S }", serviceSymbol, clientId);
+
+        writer.writeDocs("ServiceName returns the full service title.");
+        writer.write("func (c $P) ServiceID() string { return $S }", serviceSymbol, clientTitle);
```

Comment:
Nit, the doc strings are on the wrong methods.

Reply:
amazing